src/pybind/mgr/mgr_util.py

```diff
@@ -335,6 +338,79 @@ def get_all_filesystems(self) -> List[str]:
+        self._mgr = mgr
+        self._cephfs_client = client or CephfsClient(mgr)
+
+    def _extract_path_component(self, path: str, index: int) -> Optional[str]:
+        """
+        Extracts a specific component from the path based on the given index.
+
+        :param path: The path in the format '/volumes/{subvolumegroup}/{subvolume}/..'
+        :param index: The index of the component to extract (1 for subvolumegroup, 2 for subvolume)
+        :return: The component at the specified index
+        """
+        parts = path.strip('/').split('/')
+        if len(parts) >= 3 and parts[0] == "volumes":
+            return parts[index]
+        return None
+
+    def _fetch_subvolumegroup_from_path(self, path: str) -> Optional[str]:
+        """
+        Extracts and returns the subvolume group name from the given path.
+
+        :param path: The path in the format '/volumes/{subvolumegroup}/{subvolume}/..'
+        :return: The subvolume group name
+        """
+        return self._extract_path_component(path, 1)
+
+    def _fetch_subvolume_from_path(self, path: str) -> Optional[str]:
+        """
+        Extracts and returns the subvolume name from the given path.
+
+        :param path: The path in the format '/volumes/{subvolumegroup}/{subvolume}/..'
+        :return: The subvolume name
+        """
+        return self._extract_path_component(path, 2)
+
+    def _manage_earmark(self, path: str, volume: str, earmark: Optional[str] = None) -> Optional[str]:
+        """
+        Manages (get or set) the earmark for a subvolume based on the provided parameters.
+
+        :param path: The path of the subvolume
+        :param volume: The volume name
+        :param earmark: The earmark to set (None if only getting the earmark)
+        :return: The earmark if getting, otherwise None
+        """
+        with open_filesystem(self._cephfs_client, volume) as fs:
+            earmark_manager = CephFSVolumeEarmarking(fs, path)
+            try:
+                if earmark is not None:
+                    earmark_manager.set_earmark(earmark)
+                    return None
+                return earmark_manager.get_earmark()
+            except EarmarkException as e:
+                logger.error(f"Failed to manage earmark: {e}")
+                return None
+
+    def get_earmark(self, path: str, volume: str) -> Optional[str]:
+        """
+        Get earmark for a subvolume.
+        """
+        return self._manage_earmark(path, volume)
+
+    def set_earmark(self, path: str, volume: str, earmark: str) -> None:
+        """
+        Set earmark for a subvolume.
+        """
+        self._manage_earmark(path, volume, earmark)
+
+    def check_earmark(self, earmark: str, top_level_scope: EarmarkTopScope) -> bool:
+        """Check if the earmark belongs to the mentioned top level scope."""
+        return earmark.startswith(f"{top_level_scope.value}.")
```

Comment:
IIRC we can't rely on there being a dot here. Let's add a parse_earmark function to src/python-common/ceph/fs/earmarking.py to avoid issues WRT having to reimplement parsing of earmarks in different modules. Something like:

```python
class EarmarkContents(NamedTuple):
    top: EarmarkTopScope
    subsections: List[str]

class EarmarkParseError(ValueError):
    pass

def parse_earmark(value: str) -> Optional[EarmarkContents]:
    """Parse an earmark value. Returns None if value is an empty string.
    Raises EarmarkParseError if top-level scope is not valid.
    Returns an EarmarkContents for valid earmark values.
    if not value:
        return None
    # ... copy most of the body from the exiting _validate_earmark function ...
    # ... reuse this call in  _validate_earmark ...
```

Reply:
Well I assumed here that it'll be a valid earmark already bcz we ensure on setting earmark that it's valid one or not so given if it's already set I think maybe it's safe to assume that it'll be valid already?

Reply:
I'd rather not assume that. These are file system properties so other users *could* tinker with them however unlikely that'd be. Plus, I don't think you can trust that there'll be a trailing dot in any case.

Reply:
Ah okay, get it! I've added parser now & update validation accordingly. Also using parser first under `check_earmark(..)`